desc/utils.py

```diff
@@ -431,6 +431,41 @@ def svd_inv_null(A):
     return Ainv, Z
 
 
+def qr_xp_null(A, b, tol=1e-10):
+    """Compute null space of a matrix and particular solution Ax=b using QR.
+
+    Parameters
+    ----------
+    A : ndarray
+        Matrix to find null space of.
+    b : ndarray
+        Right-hand side of Ax = b.
+
+    Returns
+    -------
+    x_p : ndarray
+        Particular solution to Ax = b.
+    Z : ndarray
+        Null space of A such that AZ=0 and Z.T@Z=I.
+
+    """
+    # Linear constraint matrix A is usually wide
+    # QR decomposition of A^T
+    Q, R = qr(A.T)
+    # Determine rank
+    diag = jnp.abs(jnp.diag(R))
+    rank = jnp.sum(diag > tol)
```

Comment:
should make tol a relative tolerances based on the size of the matrix or largest diagonal or something, like for svd

Also the default should probably be smaller, like 10*eps or something

Reply:
I switched to SVD again, but I changed the `svd_inv_null` to use JAX's SVD. The CPU benchmarks won't show a lot but on GPU it is very fast for consecutive runs.